ruma-api-macros/src/derive_outgoing.rs

```diff
@@ -1,78 +1,136 @@
-    let mut gen_copy = input.generics.clone();
-    let (impl_gen, ty_gen) = split_for_impl_lifetime_less(&mut gen_copy);
+            Ok(quote! {
+                #[doc = #doc]
+                #[derive(Debug, #derive_deserialize)]
+                #vis enum #incoming_ident #ty_gen { #( #vars, )* }
 
-    let struct_def = match struct_kind {
-        StructKind::Struct => quote! { { #(#fields,)* } },
-        StructKind::Tuple => quote! { ( #(#fields,)* ); },
-    };
+                impl #original_impl_gen ::ruma_api::Outgoing for #original_ident #original_ty_gen {
+                    type Incoming = #incoming_ident #impl_gen;
+                }
+            })
+        }
+        DataKind::Struct(mut fields, struct_kind) => {
+            let mut any_attribute = false;
+            for field in &mut fields {
+                if strip_lifetimes(&mut field.ty) {
+                    any_attribute = true;
+                }
+            }
+
+            let original_ident = &input.ident;
+            let (original_impl_gen, original_ty_gen, _) = input.generics.split_for_impl();
 
-    Ok(quote! {
-        #[doc = #doc]
-        #[derive(Debug, #derive_deserialize)]
-        #vis struct #incoming_ident #ty_gen #struct_def
+            if !any_attribute {
+                return Ok(impl_outgoing_with_incoming_self(
+                    original_ident,
+                    Some(original_impl_gen),
+                    Some(original_ty_gen),
+                ));
+            }
 
-        impl #original_impl_gen ::ruma_api::Outgoing for #original_ident #original_ty_gen {
-            type Incoming = #incoming_ident #impl_gen;
+            let vis = input.vis;
+            let doc = format!("'Incoming' variant of [{ty}](struct.{ty}.html).", ty = &input.ident);
+            let incoming_ident =
+                format_ident!("Incoming{}", original_ident, span = Span::call_site());
+            let mut gen_copy = input.generics.clone();
+            let (impl_gen, ty_gen) = split_for_impl_lifetime_less(&mut gen_copy);
+
+            let struct_def = match struct_kind {
+                StructKind::Struct => quote! { { #(#fields,)* } },
+                StructKind::Tuple => quote! { ( #(#fields,)* ); },
+            };
+
+            Ok(quote! {
+                #[doc = #doc]
+                #[derive(Debug, #derive_deserialize)]
+                #vis struct #incoming_ident #ty_gen #struct_def
+
+                impl #original_impl_gen ::ruma_api::Outgoing for #original_ident #original_ty_gen {
+                    type Incoming = #incoming_ident #impl_gen;
+                }
+            })
         }
-    })
+    }
 }
 
 fn no_deserialize_in_attrs(attrs: &[Attribute]) -> bool {
     attrs.iter().any(|attr| attr.path.is_ident("incoming_no_deserialize"))
 }
 
-fn impl_outgoing_with_incoming_self(ident: Ident) -> TokenStream {
+fn impl_outgoing_with_incoming_self(
```

Comment:
Can this only take the full generics as an argument and call `split_for_impl`, or would that be more complicated (due to duplication or sth.)?

Reply:
It would duplicate the `split_for_impl` but it would actually be less lines of code (passing the args takes up more than one duplication of `split_for_impl`).

Reply:
You do end up having to clone the whole `input.data` at the begining.